include/dlaf/eigensolver/tridiag_solver/impl.h

```diff
@@ -76,37 +76,12 @@ std::vector<pika::shared_future<T>> cuppensDecomposition(Matrix<T, D>& mat_trd)
   offdiag_vals.reserve(to_sizet(i_end));
 
   for (SizeType i_split = 0; i_split < i_end; ++i_split) {
-    // Cuppen's decomposition
-    //
-    // Substracts the offdiagonal element at the split from the top and bottom diagonal elements and
-    // returns the offdiagonal element. The split is between the last row of the top tile and the first
-    // row of the bottom tile.
-    //
-    auto cuppens_fn = [](const auto& top, const auto& bottom, [[maybe_unused]] auto&&... ts) {
-      TileElementIndex offdiag_idx{top.size().rows() - 1, 1};
-      TileElementIndex top_idx{top.size().rows() - 1, 0};
-      TileElementIndex bottom_idx{0, 0};
-      if constexpr (D == Device::CPU) {
-        const T offdiag_val = top(offdiag_idx);
-
-        // Refence: Lapack working notes: LAWN 69, Serial Cuppen algorithm, Chapter 3
-        //
-        top(top_idx) -= std::abs(offdiag_val);
-        bottom(bottom_idx) -= std::abs(offdiag_val);
-        return offdiag_val;
-      }
-      else {
-        return cuppensDecompOnDevice(top.ptr(offdiag_idx), top.ptr(top_idx), bottom.ptr(bottom_idx),
-                                     ts...);
-      }
-    };
-
     auto sender =
         ex::when_all(mat_trd(LocalTileIndex(i_split, 0)), mat_trd(LocalTileIndex(i_split + 1, 0)));
 
     // Cuppen's tridiagonal decomposition
     offdiag_vals.push_back(
-        di::transform(di::Policy<DefaultBackend<D>::value>(), std::move(cuppens_fn), std::move(sender)) |
+        di::transform(di::Policy<DefaultBackend<D>::value>(), cuppensDecomp_o, std::move(sender)) |
```

Comment:
I just want to point out that this can be further reduced to `std::move(sender) | cuppensDecomp(di::Policy<DefaultBackend<D>::value>())` if you use `DLAF_MAKE_SENDER_ALGORITHM_OVERLOADS`. However, this change would be part of https://github.com/eth-cscs/DLA-Future/issues/634 (I'll link to this comment there).

Reply:
Yes, thanks for bringing this up here, I meant to talk about it in the weekly meeting. I actually intentionally left out the `DLAF_MAKE_SENDER_ALGORITHM_OVERLOADS` overloads because I wasn't sure if they add much value compared to just passing `_o` to `transform()`. Unless I am missing something, they seem to be mostly syntactic sugar which, unfortunately, adds an additional templating layer and more boilerplate. However, if there are good reasons to use them (be it just for the sake of convention) I wouldn't mind using them here.

Reply:
You're right about it being syntactic sugar, but it's not only that. Let me try to explain a few points.

1. The current set of macros definitely add bloat (and worse, increase compile times). This is bad, no doubt about it!
2. Your comment makes me think of a way to avoid some of the above. The current layering with `DLAF_MAKE_SENDER_ALGORITHM_OVERLOADS` going through a layer of `transform` is one way of implementing it but it need not be that way. The sender overloads can instead "inline" the body of `transform` directly. Since `transform` is anyway instantiated for each different function type and policy combination, this should not add bloat and hopefully should even reduce it.
3. There's a design reason for preferring `algo_name(...)` to be the entry point rather than `transform(..., algo_name)`. The former is much more straightforward to customize, and the latter is a bit sketchy to customize (and will not be possible to customize in some cases).

The two first points are relevant to https://github.com/pika-org/pika/issues/433 and I will try something out there to see if it helps.

The third point has the added dimension that there won't be the same concerns and constraints for all algorithms. For `tile::gemm` we eventually want to be able to use `std::blas::matrix_matrix_multiplication` and making this transition easier was the main intention for me to prefer the "non-transform" versions. However, we also have things like `cuppensDecomp` which will probably never be an algorithm in the standard library and where we control the implementation from top to bottom. There we're of course much more free to make various assumptions based on the implementation (such as removing a layer of function calls). If 2. can help remove the unnecessary bloat that is currently there I would always prefer using non-transform versions for consistency. If that doesn't help enough then we absolutely targeted changes in the knowledge that the changes are just that: bloat and compilation time optimizations.

Does that make sense?